### ISSUE TITLE
`vg deconstruct`  crashes with vg 1.36.0 when using a workflow that worked in vg 1.35.0

### ISSUE THREAD

**Issue by MBMumphrey** [vgteam/vg#3483]
**1. What were you trying to do?**
I was re-running my current workflow (mentioned in issue #3469) to see if the issue was resolved with vg 1.36.0. This workflow creates a graph from GFA w-lines, uses `deconstruct -a -g` to create a reference panel of all variants, and then `vg call -a -g` to genotype all variants in the reference panel. The toy example listed in issue #3469 still runs in 1.36.0, however when running the workflow on my real data, `vg deconstruct -a -g` crashes despite working when using 1.35.0. 

**2. What did you want to happen?**
To create a VCF containing all snarls in the graph using `vg deconstruct -a -g`

**3. What actually happened?**
VG crashed.

**4. If you got a line like `Stack trace path: /somewhere/on/your/computer/stacktrace.txt`, please copy-paste the contents of that file here:**

```
Crash report for vg v1.36.0 "Cibottola"
Stack trace (most recent call last) in thread 27330:
#9    Object "", at 0xffffffffffffffff, in 
#8    Object "/usr/bin/vg", at 0x1cf7692, in __clone
#7    Object "/usr/bin/vg", at 0x1336ff8, in start_thread
      Source "/build/glibc-eX1tMB/glibc-2.31/nptl/pthread_create.c", line 477, in start_thread [0x1336ff8]
#6    Object "/usr/bin/vg", at 0x1c3077d, in gomp_thread_start
#5    Object "/usr/bin/vg", at 0xd6e6b6, in vg::Deconstructor::deconstruct(std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, handlegraph::PathPositionHandleGraph const*, vg::SnarlManager*, bool, int, bool, int, bool, bool, std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, int>, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, int> > > > const*, std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, int, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, int> > > const*, gbwt::GBWT*, std::unordered_map<long long, std::pair<long long, unsigned long>, std::hash<long long>, std::equal_to<long long>, std::allocator<std::pair<long long const, std::pair<long long, unsigned long> > > > const*) [clone ._omp_fn.0]
      Source "src/deconstructor.cpp", line 932, in _ZN2vg13Deconstructor11deconstructESt6vectorINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESaIS7_EEPKN11handlegraph23PathPositionHandleGraphEPNS_12SnarlManagerEbibibbPKSt13unordered_mapIS7_St4pairIS7_iESt4hashIS7_ESt8equal_toIS7_ESaISH_IKS7_SI_EEEPKSG_IS7_iSK_SM_SaISH_ISN_iEEEPN4gbwt4GBWTEPKSG_IxSH_IxmESJ_IxESL_IxESaISH_IKxS11_EEE._omp_fn.0 [0xd6e6b6]
#4    Object "/usr/bin/vg", at 0xd6d710, in vg::Deconstructor::deconstruct_site(vg::Snarl const*) const
      Source "src/deconstructor.cpp", line 649, in deconstruct_site [0xd6d710]
#3    Object "/usr/bin/vg", at 0x1c272b5, in GOMP_parallel
#2    Object "/usr/bin/vg", at 0xd6c65a, in vg::Deconstructor::deconstruct_site(vg::Snarl const*) const [clone ._omp_fn.2]
      Source "src/deconstructor.cpp", line 678, in _ZNK2vg13Deconstructor16deconstruct_siteEPKNS_5SnarlE._omp_fn.2 [0xd6c65a]
#1    Object "/usr/bin/vg", at 0xd6b4d1, in vg::Deconstructor::get_gbwt_path_position(vg::SnarlTraversal const&, unsigned long const&) const
    | Source "src/deconstructor.cpp", line 1061, in put
      Source "/home/stephen/Documents/projects/cgl/vg_static_binary/vg/include/lru_cache.h", line 118, in get_gbwt_path_position [0xd6b4d1]
#0    Object "/usr/bin/vg", at 0xd75cf3, in LRUCache<unsigned long, std::shared_ptr<std::unordered_map<handlegraph::handle_t, unsigned long, std::hash<handlegraph::handle_t>, std::equal_to<handlegraph::handle_t>, std::allocator<std::pair<handlegraph::handle_t const, unsigned long> > > > >::Impl::put(unsigned long, std::shared_ptr<std::unordered_map<handlegraph::handle_t, unsigned long, std::hash<handlegraph::handle_t>, std::equal_to<handlegraph::handle_t>, std::allocator<std::pair<handlegraph::handle_t const, unsigned long> > > >)
    | Source "/home/stephen/Documents/projects/cgl/vg_static_binary/vg/include/lru_cache.h", line 157, in bump_to_front
      Source "/home/stephen/Documents/projects/cgl/vg_static_binary/vg/include/lru_cache.h", line 263, in put [0xd75cf3]

```

**5. What data and command can the vg dev team use to make the problem happen?**

I can provide the underlying graph and related files if needed. This command works with vg 1.35.0, but crashes with vg 1.36.0, when run on the same input files:
```
vg deconstruct graph.xg \
    -a \
    -g graph.gbwt \
    -t 12 > graph.vcf
```

**6. What does running `vg version` say?**

```
vg version v1.36.0 "Cibottola"
Compiled with g++ (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0 on Linux
Linked against libstd++ 20200808
Built by stephen@lubuntu
```

**Comment by glennhickey** [comment]
`deconstruct`, especially its threading, changed quite a bit between the last and current releases in order to accommodate cyclic references.  @ekg This looks like exactly the error you were grappling with at some point, right?  We'll have to figure out a way to fix, but if you have the patience, running with `-t 1` should work in the meantime.

**Comment by ekg** [comment]
For jemalloc to work correctly, the system needs the right VM overcommit
settings. Need to look up the relevant conversations. Adam might have the
details handy.

**Comment by adamnovak** [comment]
If the overcommit setting were wrong, `vg` should have printed a warning:

```
vg [warning]: System's vm.overcommit_memory setting is 2 (never overcommit). vg does not work well under these conditions; you may appear to run out of memory with plenty of memory left. Attempting to unsafely reconfigure jemalloc to deal better with this situation.
```

If you didn't see that @MBMumphrey, or if `cat /proc/sys/vm/overcommit_memory` is not `2`, then that's not the problem. See the conversation at https://github.com/vgteam/vg/pull/3416#issuecomment-932640660.

I think to fix this we'd need to get the full data set, and maybe have @ekg go in with the debugger, since he did the deconstruct refactor.

**Comment by MBMumphrey** [comment]
I did not get the overcommit warning, and when I check `cat /proc/sys/vm/overcommit_memory` the value is `0`. Running the command with `-t 1` as suggested by Glenn did work though. I would be happy to provide the `.xg` and `.gbwt` files if it would be helpful.

**Comment by adamnovak** [comment]
It would be good to get those; it does sound like we have a different multithreading bug here. It could be not enough caches are being allocated for all the threads that run in the actual parallel sections, or it could be some other memory corruption that is manifesting here.

**Comment by adamnovak** [comment]
OK, I went to look at this and our problem is that we have nested OMP parallel regions. `deconstruct()` runs `deconstruct_site()` in a parallel for loop, and `deconstruct_site()` runs `get_context()` in another parallel for loop. This is allowed, except we use `omp_get_thread_num()` in `get_context()`, expecting a unique value per thread, and actually all our threads are getting the same value because it gets our number in the innermost team, not a global thread number out of all OMP threads in existence.

A bunch of our code assumes this will never happen and that `omp_get_thread_num()` is basically a stable identifier out of all threads in the problem being worked on. I'm a bit worried we'll get into trouble if user code ever throws nested OMP parallelism at our parallel algorithms.

I think if we did `omp_set_nested(false)` it might work?
